hudi-client/hudi-client-common/src/main/java/org/apache/hudi/metadata/HoodieBackedTableMetadataWriter.java

```diff
@@ -557,32 +558,33 @@ private HoodieData<HoodieRecord> readRecordKeysFromBaseFiles(HoodieEngineContext
 
       final String fileId = FSUtils.getFileId(filename);
       final String instantTime = FSUtils.getCommitTime(filename);
-      HoodieFileReader reader = HoodieFileReaderFactory.getReaderFactory(HoodieRecord.HoodieRecordType.AVRO).getFileReader(hadoopConf.get(), dataFilePath);
-      Iterator<String> recordKeyIterator = reader.getRecordKeyIterator();
+      try (HoodieFileReader reader = HoodieFileReaderFactory.getReaderFactory(HoodieRecord.HoodieRecordType.AVRO).getFileReader(hadoopConf.get(), dataFilePath)) {
```

Comment:
If you include this in the try block then the reader will get closed after the try block before the returned iterator has been completely consumed by the engineContext.parallelize call.

In other words, this lambda returns an iterator which will be consumed (iterated over) outside this function call. The returned iterator holds a reference to the reader which should not be closed until the iterator has completed.

Reply:
Got you. Thanks for the clarification.